app/models/payment_intent.rb

```diff
@@ -0,0 +1,153 @@
+
+  # TODO: Refactor this or move it into this class
+  delegate :stripe_id, :stripe_status, :parameters, :money_amount, :find_account_id, to: :payment_record
+
+  # Stripe secrets are case-sensitive. Make sure that this information is not lost during encryption.
+  encrypts :client_secret, downcase: false
+
+  serialize :error_details, coder: JSON
+
+  enum wca_status: {
+    created: 'created', # A record has been created on the payment provider's system (no payment action yet initiated by the user)
+    pending: 'pending', # The user is now attempting to pay, but has not reached a completion state yet
+    partial: 'partial', # Some but not all funds have been paid (eg, if the user has selected an instalment-based payment option)
+    processing: 'processing', # The payment is in progress, but the user cannot do anything to advance its state (eg, provider waiting for outcome of bank transfer)
+    failed: 'failed', # The payment did not succeed for any reason (insufficient funds, user error) but the user can still try to pay
+    succeeded: 'succeeded', # Completion state - The full amount due is confirmed as being paid by the payment provider
+    canceled: 'canceled', # Completion state - the user has indicated that they will no longer attempt to complete payment
+  }
+
+  def started?
+    self.wca_status != 'created'
+  end
+
+  def retrieve_intent
+    if payment_record_type == 'StripeRecord'
+      Stripe::PaymentIntent.retrieve(
+        self.stripe_id,
+        client_secret: client_secret,
+        stripe_account: self.find_account_id,
+      )
+    else
+      raise "Trying to call retrieve_intent for a PaymentIntent with unmatched payment_record_type of: #{payment_record_type}"
+    end
+  end
+
+  def update_status_and_charges(api_intent, action_source, source_datetime = DateTime.current, &block)
+    if payment_record_type == 'StripeRecord'
+      update_stripe_status_and_charges(api_intent, action_source, source_datetime, &block)
+    else
+      raise "Trying to update status and charges for a PaymentIntent with unmatched payment_record_type of: #{payment_record_type}"
+    end
+  end
+
+  def update_stripe_status_and_charges(api_intent, action_source, source_datetime)
+    ActiveRecord::Base.transaction do
+      self.update!(error_details: api_intent.last_payment_error)
+      self.payment_record.assign_status(api_intent) # Will also update the wca_status
+
+      # Payment Intent lifecycle as per https://stripe.com/docs/payments/intents#intent-statuses
+      case api_intent.status
+      when 'succeeded'
+        # The payment didn't need any additional actions and is completed!
+
+        # Record the success timestamp if not already done
+        unless self.succeeded?
+          self.assign_attributes(
+            confirmed_at: source_datetime,
+            confirmation_source: action_source,
+            wca_status: payment_record.determine_wca_status,
+          )
+          save
+          payment_record.save
```

Comment:
Why?

Reply:
**Why am I calling `save` on `payment_record` if I'm already calling `save` on the `payment_intent`:** I _thought_ saving the PI would cause its related objects to be saved as well, but that's not the case - commenting out this line and running `bin/spring rspec spec/requests/registrations_controller.rb` causes the tests to fail because the updated status is not persisted to the db.

Not sure why - perhaps the polymorphic association? 

**Why save the payment_record here and not somewhere else:** We have to update the attributes of both the PI and payment_record before persisting it to the db, meaning either the PI or the payment_record will have a `save` not visible in their model code - it makes sense to me to have it happen in the model that's "pushing" the status change

Reply:
(Even changing to `update!` doesn't solve it - I still have to save the payment_record separately)

Reply:
That last point is weird, `update!` most definitely fires an SQL call. Have you looked at https://github.com/thewca/worldcubeassociation.org/pull/9073#discussion_r1548026242?

Reply:
Just staring at the code, I feel like reverting `def assign_status` to the old `def update_status` should definitely solve the problem (which is exactly what my other comment points to).

Have you specifically tried that? (Just double-checking because you did not reference my other comment)

Reply:
If all else fails, you can also set `autosave: true` in the `belongs_to` association, but I would only consider this as a last resort.